Youtube steps Code Review

gwpantazes opened the pull request:
Pull request for the purpose of code review

Commit:
Adding Youtube Step Definitions

Adding Youtube step definitions, most of which are already done. I will be
tranferring computers so I need to put this up on github, even though
there are some errors in these steps, and there are more to write.

Commit:
Fullscreen and Theater Mode Scenarios

Fullscreen and Theater mode scenarios are now working. Had to implement waits for these slower functionality, so that clicks are followed by their own little wait until an element updates.

Also cleaned up other things, such as fixing up the waits that were sitting around doing nothing.

trammel-may commented on src/test/resource/youtubeHTML5Video.feature:
```diff
@@ -32,14 +30,23 @@ Feature: HTML5 Video Player functionality on Youtube
-    Given video is fullscreen
+    Given user presses the fullscreen button
+    And video is fullscreen
     When user presses the fullscreen button
```
because you re-use "user pressed x button" you should parameterized it and pass in button names as arguments.

trammel-may commented on src/test/java/steps/YoutubeHTML5VideoStepDefinitions.java:
```diff
@@ -0,0 +1,152 @@
+        } else if (shouldTestForFullscreenOrNot.contains("not")) {
+            // Don't want to find a fullscreen indicator, so we should get empty list
+            assertTrue("Video is indicated to be fullscreen, but should be normal size.",
+                    driver.findElements(By.className("ytp-fullscreen")).size() == 0 );
```
instead of an if else, this should be mapped

Map(String, Runnable) mapExample = new HashMap(String, Runnable)();
mapExample.put("string", () -> myAssertionMethod());

mapExample.get("string").run();

this maps the assertion method to a string. If a string is not found in the map, an error is thrown. 

Because it is a hash map, this is more performant than an if else and it reduces code bloat

trammel-may commented on src/test/java/steps/YoutubeHTML5VideoStepDefinitions.java:
```diff
@@ -0,0 +1,152 @@
+            new WebDriverWait(driver, 5).until(
+                    ExpectedConditions.not(ExpectedConditions.attributeContains(
+                            driver.findElement(By.cssSelector(".html5-video-player")),
+                            "class",
```
WebDriverWaits all use an object inside the api called a Sleeper which use thread sleeps. There are no such thing as true implicit waits inside selenium. We made a simple time loop that can be used without risk of thread lock.

trammel-may commented on src/test/java/steps/YoutubeHTML5VideoStepDefinitions.java:
```diff
@@ -0,0 +1,152 @@
+    @When("^user presses the play button$")
+    public void the_user_presses_the_play_button() throws Throwable {
+        js.executeScript(videoElement + ".play();"); //Note that this is directly using the API to play, not pressing the play button
+    }
```
interacting with the api is fine, however web testing should be as authentic to the user experience as possible. Try to find a way to interact with the player buttons directly.

trammel-may commented on src/test/java/steps/YoutubeHTML5VideoStepDefinitions.java:
```diff
@@ -0,0 +1,152 @@
+public class YoutubeHTML5VideoStepDefinitions {
+    WebDriver driver;
+    JavascriptExecutor js;
+
```
is there a reason why these have package level accessibility?

gwpantazes committed:
Fixes to Access, API calls, and removing duplicate hooks

Package level accessibility changed to private for class members in step definitions.
Commented out the BEFORE and AFTER hooks that were causing double firefox spawning.
User now maximizes window for Youtube.
Methods were reordered to follow Given > When > Then a bit better.
Changed API calls to the video player to selenium interactions with the UI buttons in Youtube.
General cleanup.
